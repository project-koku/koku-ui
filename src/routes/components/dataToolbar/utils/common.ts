import type { ToolbarLabelGroup } from '@patternfly/react-core';
import type { Query } from 'api/queries/query';
import { intl } from 'components/i18n';
import messages from 'locales/messages';
import { cloneDeep } from 'lodash';
import type { Filter } from 'routes/utils/filter';
import { awsCategoryKey, awsCategoryPrefix, orgUnitIdKey, tagKey, tagPrefix } from 'utils/props';

export interface Filters {
  [key: string]: Filter[] | { [key: string]: Filter[] };
}

export interface ToolbarChipGroupExt extends ToolbarLabelGroup {
  ariaLabelKey?: string;
  placeholderKey?: string;
  resourceKey?: string;
  selectClassName?: string; // A selector from routes/components/dataToolbar/dataToolbar.scss
  selectOptions?: ToolbarLabelGroup[];
}

export const defaultFilters = {
  [awsCategoryKey]: {},
  tag: {},
};

// Remove trailing commas -- see https://issues.redhat.com/browse/COST-3641
export const cleanInput = (value: string) => {
  return value.replace(/,*$/g, '').replace(/&*$/g, '');
};

export const getActiveFilters = query => {
  const filters = cloneDeep(defaultFilters);

  const parseFilters = (key, values, isExcludes = false) => {
    if (key.indexOf(tagPrefix) !== -1) {
      if (filters.tag[key.substring(tagPrefix.length)]) {
        filters.tag[key.substring(tagPrefix.length)] = [
          ...filters.tag[key.substring(tagPrefix.length)],
          ...getFilters(key, values, isExcludes),
        ];
      } else {
        filters.tag[key.substring(tagPrefix.length)] = getFilters(key, values, isExcludes);
      }
    } else if (key.indexOf(awsCategoryPrefix) !== -1) {
      if (filters[awsCategoryKey][key.substring(awsCategoryPrefix.length)]) {
        filters[awsCategoryKey][key.substring(awsCategoryPrefix.length)] = [
          ...filters[awsCategoryKey][key.substring(awsCategoryPrefix.length)],
          ...getFilters(key, values, isExcludes),
        ];
      } else {
        filters[awsCategoryKey][key.substring(awsCategoryPrefix.length)] = getFilters(key, values, isExcludes);
      }
    } else if (filters[key]) {
      filters[key] = [...filters[key], ...getFilters(key, values, isExcludes)];
    } else {
      filters[key] = getFilters(key, values, isExcludes);
    }
  };

  if (query?.filter_by) {
    Object.keys(query.filter_by).forEach(key => {
      const values = Array.isArray(query.filter_by[key]) ? [...query.filter_by[key]] : [query.filter_by[key]];
      parseFilters(key, values);
    });
  }
  if (query?.exclude) {
    Object.keys(query.exclude).forEach(key => {
      const values = Array.isArray(query.exclude[key]) ? [...query.exclude[key]] : [query.exclude[key]];
      parseFilters(key, values, true);
    });
  }
  return filters;
};

export const getChips = (filters: Filter[]): string[] => {
  const chips = [];
  if (filters instanceof Array) {
    filters.forEach(item => {
      const label = item.toString ? item.toString() : undefined;

      chips.push({
        key: item.value,
        node: label
          ? label
          : item.isExcludes
            ? intl.formatMessage(messages.excludeLabel, { value: item.value })
            : item.value,
      });
    });
  }
  return chips;
};

export const getDefaultCategory = (categoryOptions: ToolbarLabelGroup[], groupBy: string, query: Query) => {
  if (!categoryOptions) {
    return 'name';
  }
  if (query && query.group_by && query.group_by[orgUnitIdKey]) {
    return orgUnitIdKey;
  }
  for (const option of categoryOptions) {
    if (groupBy === option.key || (groupBy && groupBy.indexOf(tagPrefix) !== -1 && option.key === tagKey)) {
      return option.key;
    }
  }
  return categoryOptions[0].key;
};

export const getFilter = (
  filterType: string,
  filterValue: string,
  isExcludes = false,
  toString = undefined
): Filter => {
  return { type: filterType, value: filterValue, isExcludes, toString };
};

export const getFilters = (filterType: string, filterValues: string[], isExcludes = false): Filter[] => {
  return filterValues.map(value => getFilter(filterType, value, isExcludes));
};

export const hasFilters = (filters: Filters) => {
  if (filters) {
    for (const filterKey of Object.keys(filters)) {
      if (filterKey === awsCategoryKey || filterKey === tagKey) {
        for (const key of Object.keys(filters[filterKey])) {
          if (filters[filterKey][key]) {
            return true;
          }
        }
      } else {
        return true;
      }
    }
  }
  return false;
};

export const onDelete = (type: any, chip: any, currentFilters) => {
  let filter = null; // Clear all
  let filters = cloneDeep(defaultFilters);

  // Todo: workaround for https://github.com/patternfly/patternfly-react/issues/3552
  // This prevents us from using a localized string, if necessary
  let _type = type && type.key ? type.key : type;
  if (_type && _type.indexOf(tagPrefix) !== -1) {
    _type = _type.slice(tagPrefix.length);
  } else if (_type && _type.indexOf(awsCategoryPrefix) !== -1) {
    _type = _type.slice(awsCategoryPrefix.length);
  }

  if (_type) {
    const excludePrefix = intl.formatMessage(messages.excludeLabel, { value: '' });
    let id = chip && chip.key ? chip.key : chip;
    if (id && id.indexOf(excludePrefix) !== -1) {
      const isExcludes = id ? id.indexOf(excludePrefix) !== -1 : false;
      id = isExcludes ? id.slice(excludePrefix.length) : id;
    }
    if (currentFilters.tag[_type]) {
      filter = currentFilters.tag[_type].find(item => item.value === id);
    } else if (currentFilters[awsCategoryKey][_type]) {
      filter = currentFilters[awsCategoryKey][_type].find(item => item.value === id);
    } else if (currentFilters[_type]) {
      filter = (currentFilters[_type] as Filter[]).find(item => item.value === id);
    }

    const newFilters: any = cloneDeep(currentFilters);
    if (newFilters.tag[_type]) {
      // Todo: use ID
      newFilters.tag[_type] = newFilters.tag[_type].filter(item => item.value !== id);
    } else if (newFilters[awsCategoryKey][_type]) {
      newFilters[awsCategoryKey][_type] = newFilters[awsCategoryKey][_type].filter(item => item.value !== id);
    } else if (newFilters[_type]) {
      newFilters[_type] = newFilters[_type].filter(item => item.value !== id);
    }
    filters = newFilters;
  }
  return {
    filter,
    filters,
  };
};
