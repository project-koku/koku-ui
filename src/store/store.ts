import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications';
import axios from 'axios';
import { applyMiddleware, compose, createStore, DeepPartial } from 'redux';
import thunk from 'redux-thunk';

import { rootReducer, RootState } from './rootReducer';

// eslint-disable-next-line @typescript-eslint/no-unused-vars
declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__(options: any): typeof compose;
  }
}

const composeEnhancers =
  typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({ serialize: true })
    : compose;

export const middlewares = [thunk, notificationsMiddleware()];

export function configureStore(initialState: DeepPartial<RootState>) {
  const enhancer = composeEnhancers(applyMiddleware(...middlewares));
  const store = createStore(rootReducer, initialState as any, enhancer);

  axios.interceptors.response.use(null, error => {
    return Promise.reject(error);
  });

  return store;
}
