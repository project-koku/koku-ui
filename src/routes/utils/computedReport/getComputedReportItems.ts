import type { Report, ReportData, ReportItem, ReportItemValue, ReportValue } from 'api/reports/report';
import { intl } from 'components/i18n';
import messages from 'locales/messages';
import { sort, SortDirection } from 'routes/utils/sort';

import { getItemLabel } from './getItemLabel';

export interface ComputedReportValue {
  units?: string;
  value?: number | string;
}

export interface ComputedReportItemValue {
  distributed?: ReportValue;
  markup?: ReportValue;
  platformDistributed?: ReportValue;
  raw?: ReportValue;
  total?: ReportValue;
  usage?: ReportValue;
  workerUnallocatedDistributed?: ReportValue;
}

export interface ComputedReportOcpItem extends ReportItem {
  capacity?: ReportValue;
  cluster?: string;
  clusters?: string[];
  limit?: ReportValue;
  request?: ReportValue;
  usage?: ReportValue;
}

export interface ComputedReportOrgItem extends ReportItem {
  id?: string;
}

export interface ComputedReportItem extends ComputedReportOcpItem, ComputedReportOrgItem {
  cost?: ReportItemValue;
  date?: string;
  delta_percent?: number;
  delta_value?: number;
  infrastructure?: ReportItemValue;
  label?: string; // helper for item label
  source_uuid?: string;
  supplementary?: ReportItemValue;
  type?: string; // 'account' or 'organizational_unit'
}

export interface ComputedReportItemsParams<R extends Report, T extends ReportItem> {
  idKey: keyof T;
  isDateMap?: boolean;
  report: R;
  sortKey?: keyof ComputedReportItem;
  sortDirection?: SortDirection;
}

export function getComputedReportItems<R extends Report, T extends ReportItem>({
  idKey,
  isDateMap,
  report,
  sortDirection = SortDirection.asc,
  sortKey = 'date',
}: ComputedReportItemsParams<R, T>) {
  return sort(
    getUnsortedComputedReportItems<R, T>({
      idKey,
      isDateMap,
      report,
      sortDirection,
      sortKey,
    }),
    {
      key: sortKey,
      direction: sortDirection,
    }
  );
}

// For filter[categpry]=platform, all clusters are listed in the breakdown page
function getClusters(val, item?: any) {
  const clusters = val.clusters ? val.clusters : [];
  if (item && item.clusters) {
    item.clusters.forEach(cluster => {
      if (!clusters.includes(cluster)) {
        clusters.push(cluster);
      }
    });
  }
  return clusters.sort((a, b) => {
    if (a < b) {
      return -1;
    }
    if (a > b) {
      return 1;
    }
    return 0;
  });
}

function getCostData(val, key, item?: any) {
  const defaultUnits = 'USD';
  return {
    ...(val[key] &&
      val[key].distributed && {
        distributed: {
          value: val[key].distributed.value + (item && item[key]?.distributed ? item[key].distributed.value : 0),
          units: val[key].distributed.units ? val[key].distributed.units : defaultUnits,
        },
      }),
    ...(val[key] &&
      val[key].markup && {
        markup: {
          value: val[key].markup.value + (item && item[key]?.markup ? item[key].markup.value : 0),
          units: val[key].markup.units ? val[key].markup.units : defaultUnits,
        },
      }),
    ...(val[key] &&
      val[key].platform_distributed && {
        platformDistributed: {
          value:
            val[key].platform_distributed.value +
            (item && item[key]?.platform_distributed ? item[key].platform_distributed.value : 0),
          units: val[key].platform_distributed.units ? val[key].platform_distributed.units : defaultUnits,
        },
      }),
    ...(val[key] &&
      val[key].total && {
        raw: {
          value: val[key].raw.value + (item && item[key]?.raw ? item[key].raw.value : 0),
          units: val[key].raw.units ? val[key].raw.units : defaultUnits,
        },
      }),
    ...(val[key] &&
      val[key].total && {
        total: {
          value: val[key].total.value + (item && item[key]?.total ? item[key].total.value : 0),
          units: val[key].total.units ? val[key].total.units : defaultUnits,
        },
      }),
    ...(val[key] &&
      val[key].usage && {
        usage: {
          value: val[key].usage.value + (item && item[key]?.usage ? item[key].usage.value : 0),
          units: val[key].usage.units ? val[key].usage.units : defaultUnits,
        },
      }),
    ...(val[key] &&
      val[key].worker_unallocated_distributed && {
        workerUnallocatedDistributed: {
          value:
            val[key].worker_unallocated_distributed.value +
            (item && item[key]?.worker_unallocated_distributed ? item[key].worker_unallocated_distributed.value : 0),
          units: val[key].worker_unallocated_distributed.units
            ? val[key].worker_unallocated_distributed.units
            : defaultUnits,
        },
      }),
  };
}

function getUsageData(val, item?: any) {
  const defaultUnits = 'Core-Hours';
  return {
    ...(val.capacity && {
      capacity: {
        value: val.capacity.value + (item?.capacity ? item.capacity.value : 0),
        units: val.capacity.units ? val.capacity.units : defaultUnits,
        ...(val.capacity.count !== undefined && {
          count: val.capacity.count + (item?.capacity ? item.capacity.count : 0),
        }),
        ...(val.capacity.count_units && {
          count_units: val.capacity.count_units,
        }),
        ...(val.capacity.unused !== undefined && {
          unused: val.capacity.unused + (item?.capacity ? item.capacity.unused : 0),
        }),
        ...(val.capacity.unused_percent !== undefined && {
          unused_percent: val.capacity.unused_percent + (item?.capacity ? item.capacity.unused_percent : 0),
        }),
      },
    }),
    ...(val.limit && {
      limit: {
        value: val.limit.value + (item?.limit ? item.limit.value : 0),
        units: val.limit.units ? val.limit.units : defaultUnits,
      },
    }),
    ...(val.request && {
      request: {
        value: val.request.value + (item?.request ? item.request.value : 0),
        units: val.request.units ? val.request.units : defaultUnits,
        ...(val.request.unused !== undefined && {
          unused: val.request.unused + (item?.request ? item.request.unused : 0),
        }),
        ...(val.request.unused_percent !== undefined && {
          unused_percent: val.request.unused_percent + (item?.request ? item.request.unused_percent : 0),
        }),
      },
    }),
    ...(val.usage && {
      usage: {
        value: val.usage.value + (item?.usage ? item.usage.value : 0),
        units: val.usage.units ? val.usage.units : defaultUnits,
      },
    }),
  };
}

// Details pages typically use this function with filter[resolution]=monthly
export function getUnsortedComputedReportItems<R extends Report, T extends ReportItem>({
  isDateMap = false,
  report,
  idKey, // Note: The idKey must use org_entities for reports, while group_by uses org_unit_id
}: ComputedReportItemsParams<R, T>) {
  if (!report) {
    return [];
  }

  // Map<string | number, ComputedReportItem | Map<string | number, ComputedReportItem>
  const itemMap = new Map();

  const visitDataPoint = (dataPoint: ReportData) => {
    const type = dataPoint.type; // Org unit type

    if (dataPoint && dataPoint.values) {
      dataPoint.values.forEach((val: any) => {
        let id = val.id ? val.id : val[idKey];
        if (!id) {
          id = val.date;
        }

        // Ensure unique map IDs -- https://github.com/project-koku/koku-ui/issues/706
        const idSuffix = idKey !== 'date' && idKey !== 'cluster' && val.cluster ? `-${val.cluster}` : '';
        const mapId = `${id}${idSuffix}`;

        // 'clusters' will contain either the cluster alias or default cluster ID
        const classification = val.classification;
        const cluster_alias = val.clusters && val.clusters.length > 0 ? val.clusters[0] : undefined;
        const cluster = cluster_alias || val.cluster;
        const date = val.date;
        const default_project = val.default_project && val.default_project.toLowerCase() === 'true';
        const delta_percent = val.delta_percent ? val.delta_percent : 0;
        const delta_value = val.delta_value ? val.delta_value : 0;
        const source_uuid = val.source_uuid ? val.source_uuid : [];

        let label;
        if (report.meta && report.meta.others && (id === 'Other' || id === 'Others')) {
          // Add count to "Others" label
          label = intl.formatMessage(messages.chartOthers, { count: report.meta.others });
        } else {
          const itemLabelKey = getItemLabel({ report, idKey, value: val });
          if (itemLabelKey === 'org_entities' && val.alias) {
            label = val.alias;
          } else if (itemLabelKey === 'account' && val.account_alias) {
            label = val.account_alias;
          } else if (itemLabelKey === 'cluster' && cluster_alias) {
            label = cluster_alias;
          } else if (val[itemLabelKey] instanceof Object) {
            label = val[itemLabelKey].value;
          } else {
            label = val[itemLabelKey];
          }
          if (label === undefined || label.trim().length === 0) {
            label = val.alias && val.alias.trim().length > 0 ? val.alias : val[idKey];
          }
        }

        if (isDateMap) {
          const data = {
            ...getUsageData(val), // capacity, limit, request, & usage
            classification,
            cluster,
            clusters: getClusters(val),
            cost: getCostData(val, 'cost'),
            date,
            default_project,
            delta_percent,
            delta_value,
            id,
            infrastructure: getCostData(val, 'infrastructure'),
            label,
            source_uuid,
            supplementary: getCostData(val, 'supplementary'),
            type,
          };
          const item = itemMap.get(mapId);
          if (item) {
            item.set(date, data);
          } else {
            const dateMap = new Map();
            dateMap.set(date, data);
            itemMap.set(mapId, dateMap);
          }
        } else {
          const item = itemMap.get(mapId);
          if (item) {
            // When applying multiple group_by params, costs may be split between regions. We need to sum those costs
            // See https://issues.redhat.com/browse/COST-1131
            itemMap.set(mapId, {
              ...item,
              ...getUsageData(val, item), // capacity, limit, request, & usage
              classification,
              cluster,
              clusters: getClusters(val, item),
              cost: getCostData(val, 'cost', item),
              date,
              default_project,
              delta_percent,
              delta_value,
              id,
              infrastructure: getCostData(val, 'infrastructure', item),
              label,
              source_uuid,
              supplementary: getCostData(val, 'supplementary', item),
              type,
            });
          } else {
            itemMap.set(mapId, {
              ...getUsageData(val), // capacity, limit, request, & usage
              classification,
              cluster,
              clusters: getClusters(val),
              cost: getCostData(val, 'cost'),
              date,
              default_project,
              delta_percent,
              delta_value,
              id,
              infrastructure: getCostData(val, 'infrastructure'),
              label,
              source_uuid,
              supplementary: getCostData(val, 'supplementary'),
              type,
            });
          }
        }
      });
    }
    for (const key in dataPoint) {
      if (dataPoint[key] instanceof Array) {
        return dataPoint[key].forEach(visitDataPoint);
      }
    }
  };
  if (report && report.data) {
    report.data.forEach(visitDataPoint);
  }
  return Array.from(itemMap.values());
}
