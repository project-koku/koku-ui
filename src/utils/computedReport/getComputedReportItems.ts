import { Report, ReportData, ReportItem, ReportItemValue, ReportValue } from 'api/reports/report';
import { createIntlEnv } from 'components/i18n/localeEnv';
import messages from 'locales/messages';
import { sort, SortDirection } from 'utils/sort';

import { getItemLabel } from './getItemLabel';

export interface ComputedReportValue {
  units?: string;
  value?: number | string;
}

export interface ComputedReportItemValue {
  markup?: ReportValue;
  raw?: ReportValue;
  total?: ReportValue;
  usage?: ReportValue;
}

export interface ComputedReportOcpItem extends ReportItem {
  capacity?: ReportValue;
  cluster?: string;
  clusters?: string[];
  limit?: ReportValue;
  request?: ReportValue;
  usage?: ReportValue;
}

export interface ComputedReportOrgItem extends ReportItem {
  id?: string;
}

export interface ComputedReportItem extends ComputedReportOcpItem, ComputedReportOrgItem {
  cost?: ReportItemValue;
  date?: string;
  delta_percent?: number;
  delta_value?: number;
  infrastructure?: ReportItemValue;
  label?: string; // helper for item label
  source_uuid?: string;
  supplementary?: ReportItemValue;
  type?: string; // 'account' or 'organizational_unit'
}

export interface ComputedReportItemsParams<R extends Report, T extends ReportItem> {
  daily?: boolean;
  idKey: keyof T;
  report: R;
  sortKey?: keyof ComputedReportItem;
  sortDirection?: SortDirection;
}

export function getComputedReportItems<R extends Report, T extends ReportItem>({
  daily,
  idKey,
  report,
  sortDirection = SortDirection.asc,
  sortKey = 'date',
}: ComputedReportItemsParams<R, T>) {
  return sort(
    getUnsortedComputedReportItems<R, T>({
      daily,
      idKey,
      report,
      sortDirection,
      sortKey,
    }),
    {
      key: sortKey,
      direction: sortDirection,
    }
  );
}

function getCostData(val, key, item?: any) {
  return {
    markup: {
      value:
        (item && item[key] && item[key].markup ? item[key].markup.value : 0) +
        (val[key] && val[key].markup ? val[key].markup.value : 0),
      units: val && val[key] && val[key].markup ? val[key].markup.units : 'USD',
    },
    raw: {
      value:
        (item && item[key] && item[key].raw ? item[key].raw.value : 0) +
        (val[key] && val[key].raw ? val[key].raw.value : 0),
      units: val && val[key] && val[key].raw ? val[key].raw.units : 'USD',
    },
    total: {
      value:
        (item && item[key] && item[key].total ? item[key].total.value : 0) +
        (val[key] && val[key].total ? Number(val[key].total.value) : 0),
      units: val && val[key] && val[key].total ? val[key].total.units : null,
    },
    usage: {
      value:
        (item && item[key] && item[key].usage ? item[key].usage.value : 0) +
        (val[key] && val[key].usage ? Number(val[key].usage.value) : 0),
      units: val && val[key] && val[key].usage ? val[key].usage.units : null,
    },
  };
}

function getUsageData(val, item?: any) {
  return {
    capacity: {
      value: (item && item.capacity ? item.capacity.value : 0) + (val.capacity ? val.capacity.value : 0),
      units: val && val.capacity ? val.capacity.units : 'Core-Hours',
    },
    limit: {
      value: (item && item.limit ? item.limit.value : 0) + (val.limit ? val.limit.value : 0),
      units: val && val.limit ? val.limit.units : 'Core-Hours',
    },
    request: {
      value: (item && item.request ? item.request.value : 0) + (val.request ? val.request.value : 0),
      units: val && val.request ? val.request.units : 'Core-Hours',
    },
    usage: {
      value: (item && item.usage ? item.usage.value : 0) + (val.usage ? val.usage.value : 0),
      units: val && val.usage ? val.usage.units : 'Core-Hours',
    },
  };
}

// Details pages typically use this function with filter[resolution]=monthly
export function getUnsortedComputedReportItems<R extends Report, T extends ReportItem>({
  daily = false,
  report,
  idKey, // Note: The idKey must use org_entities for reports, while group_by uses org_unit_id
}: ComputedReportItemsParams<R, T>) {
  if (!report) {
    return [];
  }

  // Map<string | number, ComputedReportItem | Map<string | number, ComputedReportItem>
  const itemMap = new Map();

  const visitDataPoint = (dataPoint: ReportData) => {
    const type = dataPoint.type; // Org unit type

    if (dataPoint && dataPoint.values) {
      dataPoint.values.forEach((val: any) => {
        let id = val.id ? val.id : val[idKey];
        if (!id) {
          id = val.date;
        }

        // Ensure unique map IDs -- https://github.com/project-koku/koku-ui/issues/706
        const idSuffix = idKey !== 'date' && idKey !== 'cluster' && val.cluster ? `-${val.cluster}` : '';
        const mapId = `${id}${idSuffix}`;

        // 'clusters' will contain either the cluster alias or default cluster ID
        const cluster_alias = val.clusters && val.clusters.length > 0 ? val.clusters[0] : undefined;
        const cluster = cluster_alias || val.cluster;
        const clusters = val.clusters ? val.clusters : [];
        const date = val.date;
        const delta_percent = val.delta_percent ? val.delta_percent : 0;
        const delta_value = val.delta_value ? val.delta_value : 0;
        const source_uuid = val.source_uuid ? val.source_uuid : [];

        let label;
        if (report.meta && report.meta.others && (id === 'Other' || id === 'Others')) {
          const intl = createIntlEnv();
          // Add count to "Others" label
          label = intl.formatMessage(messages.ChartOthers, { count: report.meta.others });
        } else {
          const itemLabelKey = getItemLabel({ report, idKey, value: val });
          if (itemLabelKey === 'org_entities' && val.alias) {
            label = val.alias;
          } else if (itemLabelKey === 'account' && val.account_alias) {
            label = val.account_alias;
          } else if (itemLabelKey === 'cluster' && cluster_alias) {
            label = cluster_alias;
          } else if (val[itemLabelKey] instanceof Object) {
            label = val[itemLabelKey].value;
          } else {
            label = val[itemLabelKey];
          }
          if (label === undefined || label.trim().length === 0) {
            label = val.alias && val.alias.trim().length > 0 ? val.alias : val[idKey];
          }
        }

        if (daily) {
          const data = {
            ...getUsageData(val), // capacity, limit, request, & usage
            cluster,
            clusters,
            cost: getCostData(val, 'cost'),
            date,
            delta_percent,
            delta_value,
            id,
            infrastructure: getCostData(val, 'infrastructure'),
            label,
            source_uuid,
            supplementary: getCostData(val, 'supplementary'),
            type,
          };
          const item = itemMap.get(mapId);
          if (item) {
            item.set(date, data);
          } else {
            const dateMap = new Map();
            dateMap.set(date, data);
            itemMap.set(mapId, dateMap);
          }
        } else {
          const item = itemMap.get(mapId);
          if (item) {
            // When applying multiple group_by params, costs may be split between regions. We need to sum those costs
            // See https://issues.redhat.com/browse/COST-1131
            itemMap.set(mapId, {
              ...item,
              ...getUsageData(val, item), // capacity, limit, request, & usage
              cluster,
              clusters,
              date,
              delta_percent,
              delta_value,
              cost: getCostData(val, 'cost', item),
              id,
              infrastructure: getCostData(val, 'infrastructure', item),
              label,
              source_uuid,
              supplementary: getCostData(val, 'supplementary', item),
              type,
            });
          } else {
            itemMap.set(mapId, {
              ...getUsageData(val), // capacity, limit, request, & usage
              cluster,
              clusters,
              cost: getCostData(val, 'cost'),
              date,
              delta_percent,
              delta_value,
              id,
              infrastructure: getCostData(val, 'infrastructure'),
              label,
              source_uuid,
              supplementary: getCostData(val, 'supplementary'),
              type,
            });
          }
        }
      });
    }
    for (const key in dataPoint) {
      if (dataPoint[key] instanceof Array) {
        return dataPoint[key].forEach(visitDataPoint);
      }
    }
  };
  if (report && report.data) {
    report.data.forEach(visitDataPoint);
  }
  return Array.from(itemMap.values());
}
