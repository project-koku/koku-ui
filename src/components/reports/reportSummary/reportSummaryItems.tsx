import './reportSummaryItems.scss';

import { Skeleton } from '@patternfly/react-core';
import { Report, ReportItem } from 'api/reports/report';
import React from 'react';
import { injectIntl, WrappedComponentProps } from 'react-intl';
import { FetchStatus } from 'store/common';
import {
  ComputedReportItem,
  ComputedReportItemsParams,
  getComputedReportItems,
} from 'utils/computedReport/getComputedReportItems';
import { skeletonWidth } from 'utils/skeleton';

interface ReportSummaryItemsRenderProps {
  items: ComputedReportItem[];
}

interface ReportSummaryItemsOwnProps extends ComputedReportItemsParams<Report, ReportItem> {
  children?(props: ReportSummaryItemsRenderProps): React.ReactNode;
  status?: number;
}

type ReportSummaryItemsProps = ReportSummaryItemsOwnProps & WrappedComponentProps;

class ReportSummaryItemsBase extends React.Component<ReportSummaryItemsProps> {
  public shouldComponentUpdate(nextProps: ReportSummaryItemsProps) {
    return nextProps.report !== this.props.report;
  }

  private getItems() {
    const { idKey, report } = this.props;

    const computedItems = getComputedReportItems({
      report,
      idKey,
    });

    const otherIndex = computedItems.findIndex(i => {
      const id = i.id;
      if (id && id !== null) {
        return id === 'Other' || id === 'Others';
      }
    });

    if (otherIndex !== -1) {
      return [...computedItems.slice(0, otherIndex), ...computedItems.slice(otherIndex + 1), computedItems[otherIndex]];
    }

    return computedItems;
  }

  public render() {
    const { children, status } = this.props;

    if (status === FetchStatus.inProgress) {
      return (
        <>
          <Skeleton width={skeletonWidth.md} />
          <Skeleton className="skeleton" width={skeletonWidth.md} />
          <Skeleton className="skeleton" width={skeletonWidth.md} />
          <Skeleton className="skeleton" width={skeletonWidth.md} />
        </>
      );
    } else {
      const items = this.getItems();
      return <ul>{children({ items })}</ul>;
    }
  }
}

const ReportSummaryItems = injectIntl(ReportSummaryItemsBase);

export { ReportSummaryItems, ReportSummaryItemsProps, ReportSummaryItemsRenderProps };
