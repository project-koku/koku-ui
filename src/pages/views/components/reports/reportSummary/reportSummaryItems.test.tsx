import { render, screen } from '@testing-library/react';
import React from 'react';
import { FetchStatus } from 'store/common';
import * as utils from 'utils/computedReport/getComputedReportItems';

import { ReportSummaryItems, ReportSummaryItemsProps } from './reportSummaryItems';

jest.spyOn(utils, 'getComputedReportItems');

const props: ReportSummaryItemsProps = {
  intl: {
    formatMessage: jest.fn((m, v) => JSON.stringify(v)),
  } as any,
  children: jest.fn(() => 'children'),
  status: FetchStatus.inProgress,
};

test('contains skeleton readers if in progress', () => {
  const view = render(<ReportSummaryItems {...props} />);
  expect(view.container).toMatchSnapshot();
});

test('does not update if the report is unchanged', () => {
  render(<ReportSummaryItems {...props} status={FetchStatus.complete} />);
  expect(screen.getByText(/children/i)).not.toBeNull();
});
