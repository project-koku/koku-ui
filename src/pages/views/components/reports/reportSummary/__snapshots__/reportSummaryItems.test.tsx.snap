// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contains skeleton readers if in progress 1`] = `
<div>
  <div
    class="pf-c-skeleton"
    style="--pf-c-skeleton--Width: 66%;"
  >
    <span
      class="pf-u-screen-reader"
    />
  </div>
  <div
    class="pf-c-skeleton skeleton"
    style="--pf-c-skeleton--Width: 66%;"
  >
    <span
      class="pf-u-screen-reader"
    />
  </div>
  <div
    class="pf-c-skeleton skeleton"
    style="--pf-c-skeleton--Width: 66%;"
  >
    <span
      class="pf-u-screen-reader"
    />
  </div>
  <div
    class="pf-c-skeleton skeleton"
    style="--pf-c-skeleton--Width: 66%;"
  >
    <span
      class="pf-u-screen-reader"
    />
  </div>
</div>
`;
